prometheus_client/openmetrics/exposition.py

```diff
@@ -94,24 +108,114 @@ def generate_latest(registry):
     return ''.join(output).encode('utf-8')
 
 
-def escape_metric_name(s: str) -> str:
+def escape_metric_name(s: str, escaping: str) -> str:
     """Escapes the metric name and puts it in quotes iff the name does not
     conform to the legacy Prometheus character set.
     """
-    if _is_valid_legacy_metric_name(s):
+    if len(s) == 0:
         return s
-    return '"{}"'.format(_escape(s))
+    if escaping == ALLOWUTF8:
+        if not _is_valid_legacy_metric_name(s):
+            return '"{}"'.format(_escape(s, escaping, False))
+        return _escape(s, escaping, False)
+    elif escaping == UNDERSCORES:
+        if _is_valid_legacy_metric_name(s):
+            return s
+        return _escape(s, escaping, False)
+    elif escaping == DOTS:
+        return _escape(s, escaping, False)
+    elif escaping == VALUES:
+        if _is_valid_legacy_metric_name(s):
+            return s
+        return _escape(s, escaping, False)
+    return s
 
 
-def escape_label_name(s: str) -> str:
+def escape_label_name(s: str, escaping: str) -> str:
     """Escapes the label name and puts it in quotes iff the name does not
     conform to the legacy Prometheus character set.
     """
-    if _is_valid_legacy_labelname(s):
+    if len(s) == 0:
         return s
-    return '"{}"'.format(_escape(s))
+    if escaping == ALLOWUTF8:
+        if not _is_valid_legacy_labelname(s):
+            return '"{}"'.format(_escape(s, escaping, True))
+        return _escape(s, escaping, True)
+    elif escaping == UNDERSCORES:
+        if _is_valid_legacy_labelname(s):
+            return s
+        return _escape(s, escaping, True)
+    elif escaping == DOTS:
+        return _escape(s, escaping, True)
+    elif escaping == VALUES:
+        if _is_valid_legacy_labelname(s):
+            return s
+        return _escape(s, escaping, True)
+    return s
 
 
-def _escape(s: str) -> str:
+def _escape(s: str, escaping: str, is_labelname: bool) -> str:
```

Comment:
I don't love how messy this function signature is getting...

Reply:
Yeah, it is internal only so not the end of the world. Would it help at all to have a second function for specifically escaping label names?

Reply:
I tried another approach that looks better